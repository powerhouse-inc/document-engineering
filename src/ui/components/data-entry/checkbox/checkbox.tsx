import { cn } from '../../../../scalars/lib/utils.js'
import React, { useId } from 'react'
import { FormLabel } from '../../../../scalars/components/fragments/form-label/index.js'
import { FormMessageList } from '../../../../scalars/components/fragments/form-message/index.js'
import type { InputBaseProps, WithDifference } from '../../../../scalars/components/types.js'
import { CheckboxBase, type CheckboxValue } from './checkbox-base.js'
import { CheckboxDiff } from './subcomponents/checkboxdiff.js'

type CheckboxBaseProps = Omit<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  keyof InputBaseProps<CheckboxValue> | 'onChange'
> &
  Omit<WithDifference<boolean>, 'diffMode'>

export interface CheckboxProps extends CheckboxBaseProps, InputBaseProps<CheckboxValue> {
  onChange?: ((checked: CheckboxValue) => void) | ((checked: boolean) => void)
}

const Checkbox = React.forwardRef<HTMLButtonElement, CheckboxProps>(
  (
    {
      id: idProp,
      name,
      label,
      value,
      defaultValue,
      disabled,
      required,
      description,
      errors,
      warnings,
      onChange,
      className,
      // Diff props
      viewMode = 'edition',
      baseValue,
      ...props
    },
    ref
  ) => {
    const generatedId = useId()
    const id = idProp ?? generatedId
    const hasError = !!errors?.length

    const castValue = (value: unknown) => {
      if (value === 'true') return true
      if (value === 'false') return false
      return value
    }

    if (viewMode === 'edition') {
      return (
        <div className={cn('flex flex-col gap-2')}>
          <div className={cn('group flex items-center space-x-2', className)}>
            <CheckboxBase
              id={id}
              name={name}
              checked={castValue(value ?? defaultValue) as CheckboxValue}
              disabled={disabled}
              onCheckedChange={onChange}
              required={required}
              invalid={hasError}
              aria-invalid={hasError}
              ref={ref}
              {...props}
            />
            <FormLabel
              htmlFor={id}
              required={required}
              disabled={disabled}
              hasError={hasError}
              description={description}
              className={cn(!disabled && 'group-hover:cursor-pointer')}
              inline
            >
              {label}
            </FormLabel>
          </div>
          {warnings && <FormMessageList type="warning" messages={warnings} />}
          {errors && <FormMessageList type="error" messages={errors} />}
        </div>
      )
    }
    return (
      <CheckboxDiff
        value={value as boolean}
        label={label}
        required={required}
        disabled={disabled}
        baseValue={baseValue}
        viewMode={viewMode}
      />
    )
  }
)

Checkbox.displayName = 'Checkbox'

export { Checkbox }
