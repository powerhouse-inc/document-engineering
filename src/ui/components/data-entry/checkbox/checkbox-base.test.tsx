import { fireEvent, render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { describe, expect, it, vi } from 'vitest'
import { CheckboxBase } from './checkbox-base.js'
import { Checkbox } from './checkbox.js'

describe('Checkbox Component', () => {
  it('should match snapshot', () => {
    const { container } = render(<CheckboxBase name="test" />)
    expect(container).toMatchSnapshot()
  })

  it('should render a checkbox', () => {
    render(<CheckboxBase />)
    expect(screen.getByRole('checkbox')).toBeInTheDocument()
  })

  it('should handle checked state', () => {
    const handleCheckedChange = vi.fn()
    render(<CheckboxBase onCheckedChange={handleCheckedChange} />)

    const checkbox = screen.getByRole('checkbox')
    fireEvent.click(checkbox)

    expect(handleCheckedChange).toHaveBeenCalledWith(true)
  })

  it('should be disabled when disabled prop is true', () => {
    render(<CheckboxBase disabled />)
    expect(screen.getByRole('checkbox')).toBeDisabled()
  })

  it('should be focusable with tab navigation', async () => {
    const user = userEvent.setup()
    render(<CheckboxBase />)

    const checkbox = screen.getByRole('checkbox')
    await user.tab()

    expect(checkbox).toHaveFocus()
  })

  it('should have aria-required attribute when required', () => {
    render(<CheckboxBase required />)
    expect(screen.getByRole('checkbox')).toHaveAttribute('aria-required', 'true')
  })
})

describe('Checkbox differences', () => {
  it('should show CheckboxDiff when viewMode is addition', () => {
    render(<Checkbox viewMode="addition" value={true} baseValue={false} label="Checkbox" />)

    expect(screen.getByTestId('checkbox-diff')).toBeInTheDocument()
    expect(screen.getByRole('checkbox')).toBeChecked()
    expect(screen.getByText('Checkbox')).toBeInTheDocument()
  })
  it('should show CheckboxDiff when viewMode is removal', () => {
    render(<Checkbox viewMode="removal" value={false} baseValue={true} label="Checkbox" />)

    expect(screen.getByTestId('checkbox-diff')).toBeInTheDocument()
    expect(screen.getByRole('checkbox')).not.toBeChecked()
    expect(screen.getByText('Checkbox')).toBeInTheDocument()
  })
  it('should not render CheckboxDiff when viewMode is edition', () => {
    render(<Checkbox viewMode="edition" value={true} label="Checkbox" />)
    expect(screen.queryByTestId('checkbox-diff')).not.toBeInTheDocument()
  })
})
