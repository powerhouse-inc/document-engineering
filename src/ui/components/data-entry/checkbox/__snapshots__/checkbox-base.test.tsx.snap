// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox Component > should match snapshot 1`] = `
<div>
  <button
    aria-checked="false"
    class="peer size-4 shrink-0 rounded border-input border shadow-sm shadow-black/[.04] ring-offset-background transition-shadow focus-visible:border-ring focus-visible:ring-ring focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:border-gray-700 disabled:data-[invalid=false]:data-[state=checked]:bg-gray-700 disabled:data-[invalid=false]:data-[state=indeterminate]:bg-gray-700 dark:disabled:data-[invalid=false]:data-[state=checked]:bg-gray-500 dark:disabled:data-[invalid=false]:data-[state=indeterminate]:bg-gray-500 data-[state]:border-gray-700 dark:data-[state]:border-gray-500 data-[state=checked]:bg-gray-900 data-[state=indeterminate]:bg-gray-900 dark:data-[state=checked]:bg-gray-400 dark:data-[state=indeterminate]:bg-gray-400 data-[state=checked]:text-slate-50 data-[state=indeterminate]:text-slate-50 dark:data-[state=checked]:text-gray-900 dark:data-[state=indeterminate]:text-gray-900 group-hover:border-gray-900 data-[state=checked]:group-hover:bg-gray-900 data-[state=indeterminate]:group-hover:bg-gray-900 dark:group-hover:border-slate-50 dark:data-[state=checked]:group-hover:bg-slate-50 dark:data-[state=indeterminate]:group-hover:bg-slate-50 data-[invalid=true]:data-[state]:!border-red-800 data-[invalid=true]:data-[state=checked]:!bg-red-800 data-[invalid=true]:data-[state=indeterminate]:!bg-red-800 dark:data-[invalid=true]:data-[state]:!border-red-800 dark:data-[invalid=true]:data-[state=checked]:!bg-red-800 dark:data-[invalid=true]:data-[state=indeterminate]:!bg-red-800 data-[invalid=true]:group-hover:!border-red-900 data-[invalid=true]:data-[state=checked]:group-hover:!bg-red-900 data-[invalid=true]:data-[state=indeterminate]:group-hover:!bg-red-900"
    data-state="unchecked"
    role="checkbox"
    type="button"
    value="on"
  />
</div>
`;
